### PR TITLE
Update for changes to use George Steed's subtyping model

### PR DESCRIPTION
**Pull request by jasoncarr0**
First pass at updated tutorial for  https://github.com/ponylang/ponyc/pull/3643

This updates the subtyping, but also importantly changes how certain checks are presented. For instance, when trying to assign a `iso` expression to an `trn` variable without consuming, we will get:

`Bar iso is not a subtype of Bar trn^`

whereas before we would get:

`Bar iso! is not a subtype of Bar trn`

So there are some updates to how we talk about aliasing. There may also need to be some organization improvements, but this is mostly remaining consistent in when we talk about ephemeral and aliased caps.

Fixes/makes obsolete:  #379 #322 #52

### CODE REVIEW

File: content/reference-capabilities/capability-subtyping.md
Comment (jemc) [inline content/reference-capabilities/capability-subtyping.md:15]: Small suggestion to the first sentence to make it clearer that these four aren't the entire set of capabilities available.

```suggestion
First, let's look at the four capabilities `ref`, `val`, `box`, and `tag`. These have a very useful property: they alias to themselves (and unalias to themselves, as well). This will make the subtyping a lot simpler to work with. Then we can talk about `iso` and `trn`.
```

Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md]: Good point, applied.

**Review by jemc — APPROVED** [review]
This is looking really good! Great way of explaining it all, in my opinion.

**Review by rhagenson — CHANGES_REQUESTED** [review]
Please expand words like "cap" to their full form "capability" and "subcap" to "subcapability".

Also, as per previous reviews I have done for the Tutorial, please expand contractions.

There are some parts I think are more technical-forward and for that I can only speak from my process. After I write a technical piece I will read through it and find instances where I semantically wrote "In other words, ..." because the second explain is often easier to understand. I will then either keep only the second explaination, or flip the "In other words, ..." To be first then make the technical details a "That means, ..." or "In more detail, ..." that helps make the reading more approachable while also not removing important details if the second explaination was detailed enough by itself.

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md:33]: Good catch, do you have any good ideas for alternatives? Maybe just immediately?

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md:46]: The goal is to remind the reader that there are some subtyping relations which aren't just these enumerated instances, but that derive from them using transitivity.

File: content/reference-capabilities/capability-subtyping.md
Comment (rhagenson) [reply content/reference-capabilities/capability-subtyping.md:21]: Realized the formatting on this hide part of the comment. I think my meaning was understood though.

I did my initial review during a virtual meeting that I only needed to listen in on so apologies for that formatting mistake.

File: content/reference-capabilities/capability-subtyping.md
Comment (rhagenson) [reply content/reference-capabilities/capability-subtyping.md:33]: Good question. The concept being expressed here is that of `!` and how it interacts with `iso` and `trn` so either:

> First thing of importance: `iso! = tag` and `trn! = box`...

Or you could take a topic and stress positions focused approach, so shift the paragraph around to something like:

> Let us review the importance of `!` for capabilities. Remember that `!` is the modifier for stable, named aliases; so using it gives the strongest capability that is compatible with the original capability. In the case of `iso`, nothing else is allowed access, so with `iso!` we get `tag`, while for `trn` we allow readable aliases, so with `trn!` we get `box`. Meaning, `iso! = tag`, and `trn! = box`.

---

If you are wondering where I get what can sometimes seem like arbitrary writing rules, it is a combination of experience writing with/for Malagasy (Madagascar nationals) and following guides such as: https://www.springernature.com/gp/authors/campaigns/writing-in-english

File: content/reference-capabilities/capability-subtyping.md
Comment (rhagenson) [reply content/reference-capabilities/capability-subtyping.md:46]: Ah, I see. That makes sense now that you explain it, but I did not understand it on first read through.

How does this sound to you?

> All other capability subtype relationships can be derived transitively from all instances mentioned above. Much like we did above to determine `val <: tag` based on `val <: box` and `box <: tag`.

I don't think that improves it greatly, but it clarifies what "everything else" was and provides a brief (but perhaps unnecessary) how transitivity work for capabilities.

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md:33]: Thanks for the thoughts here. I'm thinking I might need to go and elaborate more on this in previous chapters as well; previously this subtyping chapter played a double role of explaining some aliasing and unaliasing as well as the subtyping relation itself. So in a bit of a winding way, I like the examples you've provided, but I'm thinking perhaps a bit of restructuring will be needed here. This has scaled up a bit past the idiom but the original reason I was writing that this way was related to pacing.

> If you are wondering where I get what can sometimes seem like arbitrary writing rules, it is a combination of experience writing with/for Malagasy (Madagascar nationals) and following guides such as: https://www.springernature.com/gp/authors/campaigns/writing-in-english

I have actually felt that the rules you've given have been sensible enough that they haven't felt arbitrary. But that resource will be quite useful to read through, many thanks for sharing

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md:46]: I'm not sure two copies of the example is necessary. I'll see if I can merge the two locations that mention it.

**Comment by jasoncarr0** [comment]
Regarding the contractions point, I made sure to eliminate any uses of `'ll` and `'ve`, but kept in can't. Maybe we could talk about this if you're not aligned, but `cannot` sounds incredibly stilted in each situation here, and it doesn't seem like `can't` is one that's likely to cause confusion (being much more common than cannot).

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md]: Applied in f0c12746216a6d2ff6662273ac41a42006ef5572

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md]: Applied except the change can't -> cannot

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md]: Applied this in 7c9a6e468dcfa8c2a45d1d424380965f605650bc, it's very parallel but not in a bad way I think.

Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md]: Although I'm not sure how much I should make the exact phrasing identical, I don't feel as though the use of requires/grants could be confusing here although it is a bit odd

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md]: So what seemed to work out here was  to add a tiny bit more in the aliasing types section to make the `!` uses make sense, then killing the aliasing bit from here since aliasing isn't really coupled to subtyping as much anymore.

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md]: I merged  these two sections into the first one in 
0d7ef8d

File: content/reference-capabilities/capability-subtyping.md
Comment (rhagenson) [inline content/reference-capabilities/capability-subtyping.md:47]: "capabillity" -> "capability"

Simple typo.

File: content/reference-capabilities/capability-subtyping.md
Comment (rhagenson) [reply content/reference-capabilities/capability-subtyping.md]: Sounds good to me.

File: content/reference-capabilities/capability-subtyping.md
Comment (rhagenson) [reply content/reference-capabilities/capability-subtyping.md]: I think moving the aliasing made it easier to understand.

**Review by rhagenson — APPROVED** [review]
I think the rearrangement works better and all of my major changes have been resolved. 

Only found a typo that should be fixed before merging so going to approve this.

Jason, thank you for your always thoughtful contributions!

File: content/reference-capabilities/capability-subtyping.md
Comment (jasoncarr0) [reply content/reference-capabilities/capability-subtyping.md:47]: Nice catch!

**Comment by schnittchen** [comment]
This PR looks stalled for no obvious reasons. Can it be merged?

**Comment by ergl** [comment]
@schnittchen The linked PR in ponyc hasn't been merged yet, this one should probably wait until that's done.

File: docs/reference-capabilities/capability-subtyping.md
Comment (ergl) [inline docs/reference-capabilities/capability-subtyping.md:36]: What's "stable" in this context (stable aliases)?

Comment (jasoncarr0) [reply docs/reference-capabilities/capability-subtyping.md:36]: A stable alias is a named reference to something. Not sure exactly what the step would be to improve it within this paragraph, but I definitely see that it needs to be changed.

**Review by jemc — APPROVED** [review]

**Comment by SeanTAllen** [comment]
I added a "DO NOT MERGE" label as this shouldn't be merged without the corresponding changes to ponyc. I'll merge this as part of the release process for the next version of ponyc.

**Pull request merged**
